state-synchronizer/src/coordinator.rs

```diff
@@ -1043,12 +1051,17 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
         // Optimistically fetch the next chunk.
         let new_version =
             self.local_state.highest_version_in_local_storage() + txn_list_with_proof.len() as u64;
-        // The epoch in the optimistic request should be the next epoch if the current chunk
+        // The epoch in the optimistic request (= new_epoch) should be the next epoch if the current chunk
         // is the last one in its epoch.
         let new_epoch = end_of_epoch_li
             .as_ref()
+            .map(|li| {
+                // verify end-of-epoch-li against local state
+                self.local_state.trusted_epoch.verify(&li).map(|_| li)
+            })
+            .transpose()?
             .map_or(self.local_state.epoch(), |li| {
-                if li.ledger_info().version() == new_version {
+                if li.ledger_info().version() == new_version && li.ledger_info().ends_epoch() {
```

Comment:
good call

Reply:
maybe we should verify the end_of_epoch_li early and add the ends_epoch check too, if it fails, it's verification error and should be logged.

Reply:
there's a balance between the optimistic request... I kinda don't wanna rock the boat more than we already are

Reply:
> maybe we should verify the end_of_epoch_li early and add the ends_epoch check too, if it fails, it's verification error and should be logged.

lol i had this before too, but asked to change for better readability. also i think in the current code, we also verify the waypoint_li after the network send too - i think the idea was to parallelize network send and sig verification, etc